File: java/leetcode_grind/Day784.java
package leetcode_grind;

public class Day784 {
    // https://leetcode.com/problems/construct-k-palindrome-strings/description/?envType=daily-question&envId=2025-01-11
    static class Solution1 {
        public boolean canConstruct(String s, int k) {
            if (s.length() < k)
                return false;
            if (s.length() == k)
                return true;

            int[] freq = new int[26];
            int oddCount = 0;

            for (char chr : s.toCharArray()) {
                freq[chr - 'a']++;
            }

            for (int count : freq) {
                if (count % 2 == 1) {
                    oddCount++;
                }
            }

            return oddCount <= k;
        }
    }

}

File: src/leetcode_grind/day_784.rs
// https://leetcode.com/problems/construct-k-palindrome-strings/description/?envType=daily-question&envId=2025-01-11
pub fn can_construct(s: String, k: i32) -> bool {
    if s.len() < k as usize {
        return false;
    }
    if s.len() == k as usize {
        return true;
    }

    let mut odd_count = 0;
    for chr in s.chars() {
        odd_count ^= (1 << (chr as i32 - 'a' as i32));
    }
    let mut cnt = 0;
    while odd_count > 0 {
        cnt += (odd_count & 1);
        odd_count >>= 1;
    }
    cnt <= k
}
